Add BasicStation time sync protocol

adriansmares opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/3546
References https://github.com/TheThingsNetwork/lorawan-stack/issues/4804

#### Changes
<!-- What are the changes made in this pull request? -->

- Use `RefTime` for RTT calculations iff it is not `0.0`, and within the configured bound
- Add PPS based `TimeSync` support (based on the request-response format)
- Add transfer based `TimeSync` support (based on the time transfer format)
- Send `MuxTime` also during `TimeSync` ([supported as well](https://github.com/lorabasics/basicstation/blob/bd17e53ab1137de6abb5ae48d6f3d52f6c268299/src/s2e.c#L1616))

#### Testing

<!-- How did you verify that this change works? -->

Local testing with TTIGs/Laird/Tektelic `2.0.4`-based forwarders, and unit testing.

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

It seems that the Tektelic Basic Station bridge does not support the transfer-based `timesync` message:

```
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_bstn_lns.c:389 data received from WS stream:
{"msgtype":"timesync","xtime":32651101339486062,"gpstime":1319586169418490}
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [WARN  ] bridge_json_helpers.c:35 unable to retrieve field
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_bstn_lns.c:1469 stopping LWS main loop...
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_bstn_lns.c:711 disconnecting WS stream
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_bstn_lns.c:572 WS connection closed
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_bstn_lns.c:576 WSI destroyed, notifying outer loop to stop processing
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [DEBUG ] bridge_data_conv.c:1093 UDP bridge is deinitialized
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [WARN  ] main_bstn.c:162 non-critical error raised: [-1]
Oct 29 23:42:31 kona-micro tek_bstn[14117]: [WARN  ] main_bstn.c:163 restarting Basic Station in 20 seconds...
```

So this is an open problem - do we put the transfer-based `timesync` behind a feature flag ? I would be surprised if the PPS-style handshake would be used that much in the wild.

Other native gateways (TTIG / Laird) seem to handle this correctly, and RTTs are realistic.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

adriansmares committed:
gs: Record RTTs conditionally

adriansmares committed:
gs: Add TimeSync PPS-based protocol

adriansmares committed:
gs: Expose concentrator time conversion

adriansmares committed:
gs: Add TimeSync transfer-based protocol

adriansmares committed:
gs: Use RWMutex for websocket session data

adriansmares committed:
gs: Always inject session in context

adriansmares committed:
gs: Remove gateway unique ID from formatter

johanstokking commented:
> So this is an open problem - do we put the transfer-based `timesync` behind a feature flag ? I would be surprised if the PPS-style handshake would be used that much in the wild.

So the impact is that the bridge restarts?

Can we recognize timesync support by the reported version?

adriansmares commented:
> So the impact is that the bridge restarts?

Yes. 20+ seconds of downtime every 10 minutes (on the current default configuration).

> Can we recognize timesync support by the reported version?

`timesync` is not a [feature](https://doc.sm.tc/station/tcproto.html#version-message) that the bridge may or may not support - it's supposed to always be supported. With that being said, we indeed may disable `timesync` attempts based on the actual gateway version

```
  "data": {
    "@type": "type.googleapis.com/ttn.lorawan.v3.GatewayStatus",
    "time": "2021-11-01T11:21:34.449141918Z",
    "versions": {
      "firmware": "1.5.1",
      "package": "Tektelic Basic Station",
      "platform": "micro - Firmware 1.5.1 - Protocol 2",
      "station": "Tektelic Kona gateway"
    },
    "advanced": {
      "features": "",
      "model": "micro"
    }
  },
```

We can disable `timesync` for `package` = `Tektelic Basic Station` and `firmware` smaller or equal to `1.5.1`.

KrishnaIyer commented on pkg/gatewayserver/io/ws/format.go:
```diff
@@ -46,5 +46,11 @@ type Formatter interface {
+	FromDownlink(ctx context.Context, down ttnpb.DownlinkMessage, bandID string, concentratorTime scheduling.ConcentratorTime, dlTime time.Time) ([]byte, error)
+}
+
+// TimeSyncer abstracts away time transferring activities to websocket based gateways.
+type TimeSyncer interface {
+	// TransferTime generates a spurious time transfer message for a particular server time.
+	TransferTime(ctx context.Context, t time.Time, conn *io.Connection) ([]byte, error)
```
Why is this a separate interface? This is just another downstream message right?

adriansmares replied:
In order to avoid having to implementing it in other websocket frontends which do not support it. There is no point in creating a timer if the format does not support it.

Edit: Actually since we need to check at runtime anyway if the transfer should take place, I'm ok with adding this to the existing interface.

KrishnaIyer commented on pkg/gatewayserver/io/ws/lbslns/timesync.go:
```diff
@@ -0,0 +1,46 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lbslns
```
I prefer this to be a part of the `downstream` file since this is just a downstream message.

adriansmares replied:
As the logic grew a bit https://github.com/TheThingsNetwork/lorawan-stack/pull/4805/commits/8d2ca659582064642cbff8f992775a80629e2aec, I'm thinking about keeping this as a separate file. If we discard that commit, I can move the singular function to `downstream.go`.

Edit: Removing this now. Found a cleaner way.

KrishnaIyer commented on pkg/gatewayserver/io/ws/lbslns/timesync_test.go:
```diff
@@ -0,0 +1,97 @@
+
+	// No time sync available.
+	b, err := f.TransferTime(ctx, time.Now(), conn)
+	if !a.So(err, should.BeNil) {
```
Nit; `a.So(err, should.NotBeNil)` is easier to read.

adriansmares replied:
The double negation is intentional. The error _should_ be `nil`. If we use `a.So(err, should.NotBeNil)`, the test fails even though the assertion is true.

KrishnaIyer commented on pkg/gatewayserver/io/ws/lbslns/upstream.go:
```diff
@@ -581,20 +636,16 @@ func (f *lbsLNS) HandleUp(ctx context.Context, raw []byte, ids ttnpb.GatewayIden
+			return nil, err
 		}
-		recordTime(recordRTT, txConf.RefTime, txConf.XTime, receivedAt)
+		return json.Marshal(req.Response(receivedAt))
```
This message has it's own marshaler so you can use that right?

adriansmares replied:
Not a huge fan of calling the marshaller methods directly, but I can change this.

KrishnaIyer commented on pkg/gatewayserver/io/ws/lbslns/version.go:
```diff
@@ -73,7 +73,13 @@ func (f *lbsLNS) GetRouterConfig(ctx context.Context, msg []byte, bandID string,
 		return nil, nil, nil, err
 	}
-	routerCfg, err := cfg.MarshalJSON()
+	routerCfg, err := json.Marshal(struct {
+		pfconfig.RouterConfig
+		MuxTime float64 `json:"MuxTime"`
+	}{
+		RouterConfig: cfg,
+		MuxTime:      TimeToUnixSeconds(receivedAt),
+	})
```
`MuxTime` is already added here by `pfconfig` IIRC. Why is this needed?

adriansmares replied:
You're right.

KrishnaIyer reviewed (changes requested):
How about we add a `TimeSyncPeriod` to the `gateway.proto`?
1. Users can optimise this based on their needs. I think one message every 30 mins is also fine as opposed to the default of 10 mins.
2. By default, this parameter is set to zero and we disable `timesync` message.
3. Users can then enable this for their gateways and so we can get around the support per gateway brand?

There's still the issue of how to deal with current gateways. Not sure of that. For TTIGs we can run an `is-db` migration and set this to some default value.

adriansmares committed:
gs: Add session utilities

adriansmares commented:
As I've mentioned above, making this kind of behavior opt-in instead of opt-out is not something that I think we should be doing - we already have existing gateways, which are not TTIGs, that would benefit from `timesync`. We cannot control this at database level in order to migrate them:

- Some gateways have been created before the 'require secure connection' existed. Mind you - when you create a gateway via the Console, this option is not enabled by default.
- We don't have version identifiers set for most, if not all existing gateways.
- We know that existing users will hardly upgrade their existing configuration.

I've updated the synchronization interval to 30 minutes, as proposed. An implementation of the selective skipping can be found in https://github.com/TheThingsNetwork/lorawan-stack/pull/4805/commits/3fd6d133e260b41983e25985e38e655fd5a1a581 - let me know what you think.

Edit: Apparently the Tektelic bridge is even more interesting than I initially thought. They don't implement the `version` message correctly. Normally, after discovery, BasicStation connects, sends the `version` message, then starts streaming traffic.

The Tektelic bridge connects, sends the `version` message, disconnects, connects, then starts streaming traffic. I've updated the support detection code to detect if a `version` message has been sent in the current session. Then when they fix these issues it will work out of the box.

adriansmares committed:
gs: Skip timesync for unsupported versions

KrishnaIyer commented:
> I've updated the synchronization interval to 30 minutes, as proposed. An implementation of the selective skipping can be found in 3fd6d13 - let me know what you think.

Looks great to me.

johanstokking commented:
> * Some gateways have been created before the 'require secure connection' existed. Mind you - when you create a gateway via the Console, this option is not enabled by default.
> * We don't have version identifiers set for most, if not all existing gateways.
> * We know that existing users will hardly upgrade their existing configuration.

I agree with these points to not make this a per-gateway setting.

johanstokking commented on pkg/gatewayserver/io/ws/ws_test.go:
```diff
@@ -1049,6 +1059,52 @@ func TestTraffic(t *testing.T) {
+					}()
+					select {
+					case res := <-resCh:
+						switch expected := tc.ExpectedBSDownstream.(type) {
+						case lbslns.TimeSyncResponse:
```
nit; if you panic anyway you might as well just type cast normally

johanstokking reviewed (approved):
Looks really solid!

KrishnaIyer reviewed (approved):
Looks great. Let's ship it.
